oss_src/sframe_query_engine/planning/optimizations/union_transforms.hpp

```diff
@@ -99,10 +104,23 @@ class opt_union_on_source : public opt_union_transform {
           size_t begin_index = n->inputs[i]->p("begin_index");
           size_t end_index   = n->inputs[i]->p("end_index");
 
-          distinct_input_ranges.insert({begin_index, end_index});
+          size_t size = 0;
+
+          if(t == planner_node_type::SFRAME_SOURCE_NODE) {
+            // Get the size of the sframe
+            size = n->inputs[i]->any_p<sframe>("sframe").num_rows();
+          } else {
+            // Get the size of the sarray
```

Comment:
Yes, it can have dozens of types, but the others are excluded (they don't apply to this transform) by the if statement in lines 96/101 and 97/102.  Both of these have a common set of parameters, minus retrieving the size.